File: 2022/sample6.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-

import cv2
import numpy as np

img = None
def main():
    global img
    # (1)画像領域を確保し，初期化する
    img = np.zeros((480,640,3), np.uint8)

    # (2)ウィンドウを作成し，マウスイベントに対するコールバック関数を登録する
    cv2.namedWindow ("Image", cv2.WINDOW_AUTOSIZE);
    cv2.setMouseCallback ("Image", on_mouse, 0);
    cv2.imshow ("Image", img);

    while True:
        c = cv2.waitKey (2)
        if c == 27 :
            return 1;

    cv2.DestroyWindow ("Image");
    cv2.ReleaseImage (img);

# コールバック関数
def on_mouse(event, x, y, flags, param):
    global img
    # char str[64];
    max_line = 15
    w = 15
    h = 30
    str = ''
    if event == cv2.EVENT_MOUSEMOVE:
        str = "(%d,%d) %s" % (x, y, "MOUSE_MOVE")
    elif event == cv2.EVENT_LBUTTONDOWN:
        str = "(%d,%d) %s" % (x, y, "LBUTTON_DOWN")
    elif event == cv2.EVENT_RBUTTONDOWN:
        str = "(%d,%d) %s" % (x, y, "RBUTTON_DOWN")
    elif event == cv2.EVENT_MBUTTONDOWN:
        str = "(%d,%d) %s" % (x, y, "MBUTTON_DOWN")
    elif event == cv2.EVENT_LBUTTONUP:
        str = "(%d,%d) %s" % (x, y, "LBUTTON_UP")
    elif event == cv2.EVENT_RBUTTONUP:
        str = "(%d,%d) %s" % (x, y, "RBUTTON_UP")
    elif event == cv2.EVENT_MBUTTONUP:
        str = "(%d,%d) %s" % (x, y, "MBUTTON_UP")
    elif event == cv2.EVENT_LBUTTONDBLCLK:
        str = "(%d,%d) %s" % (x, y, "LBUTTON_DOUBLE_CLICK")
    elif event == cv2.EVENT_RBUTTONDBLCLK:
        str = "(%d,%d) %s" % (x, y, "RBUTTON_DOUBLE_CLICK")
    elif event == cv2.EVENT_MBUTTONDBLCLK:
        str = "(%d,%d) %s" % (x, y, "MBUTTON_DOUBLE_CLICK")

    # (5)マウスボタン，修飾キーを取得
    if (flags & cv2.EVENT_FLAG_LBUTTON):
        str += " + LBUTTON"
    elif (flags & cv2.EVENT_FLAG_RBUTTON):
        str += " + RBUTTON"
    elif (flags & cv2.EVENT_FLAG_MBUTTON):
        str += " + MBUTTON"
    elif (flags & cv2.EVENT_FLAG_CTRLKEY):
        str += " + CTRL"
    elif (flags & cv2.EVENT_FLAG_SHIFTKEY):
        str += " + SHIFT"
    elif (flags & cv2.EVENT_FLAG_ALTKEY):
        str += " + ALT"

    # (6)マウス座標，イベント，修飾キーなどを画像に描画，表示
    if on_mouse.line > max_line:
        img = cv2.getRectSubPix (img, (640,480), (320 - 0.5, 240 - 0.5 + h));
        cv2.putText (img, str, (w, 20 + h * max_line),
                     cv2.FONT_HERSHEY_DUPLEX, 0.4, (0, 200, 100));
    else:
        cv2.putText (img, str, (w, 20 + h * on_mouse.line),
                     cv2.FONT_HERSHEY_DUPLEX, 0.4, (0, 200, 100));

    on_mouse.line += 1
    cv2.imshow("Image", img);

## https://stackoverflow.com/questions/279561/
##      what-is-the-python-equivalent-of-static-variables-inside-a-function
on_mouse.line = 0;
main()

File: 2022/sample6.c
#include <cv.h>
#include <highgui.h>
#include <stdio.h>

/* グローバル変数 */
CvFont font;
IplImage *img = 0;

/* プロトタイプ宣言 */
void on_mouse (int event, int x, int y, int flags, void *param);

int
main (int argc, char *argv[])
{
  unsigned char c;

  // (1)画像領域を確保し，初期化する
  img = cvCreateImage (cvSize (640, 480), IPL_DEPTH_8U, 3);
  cvZero (img);
  cvInitFont (&font, CV_FONT_HERSHEY_DUPLEX, 0.4, 0.4, 0.0, 1, CV_AA);

  // (2)ウィンドウを作成し，マウスイベントに対するコールバック関数を登録する
  cvNamedWindow ("Image", CV_WINDOW_AUTOSIZE);
  cvSetMouseCallback ("Image", on_mouse, 0);
  cvShowImage ("Image", img);

  // (3)'Esc'キーが押された場合に終了する
  while (1) {
    c = cvWaitKey (2) & 0xff;
    if (c == '\x1b')
      return 1;
  }

  cvDestroyWindow ("Image");
  cvReleaseImage (&img);

  return 0;
}

/* コールバック関数 */
void
on_mouse (int event, int x, int y, int flags, void *param)
{
  char str[64];
  static int line = 0;
  const int max_line = 15, w = 15, h = 30;

  // (4)マウスイベントを取得
  switch (event) {
  case CV_EVENT_MOUSEMOVE:
    snprintf (str, 64, "(%d,%d) %s", x, y, "MOUSE_MOVE");
    break;
  case CV_EVENT_LBUTTONDOWN:
    snprintf (str, 64, "(%d,%d) %s", x, y, "LBUTTON_DOWN");
    break;
  case CV_EVENT_RBUTTONDOWN:
    snprintf (str, 64, "(%d,%d) %s", x, y, "RBUTTON_DOWN");
    break;
  case CV_EVENT_MBUTTONDOWN:
    snprintf (str, 64, "(%d,%d) %s", x, y, "MBUTTON_DOWN");
    break;
  case CV_EVENT_LBUTTONUP:
    snprintf (str, 64, "(%d,%d) %s", x, y, "LBUTTON_UP");
    break;
  case CV_EVENT_RBUTTONUP:
    snprintf (str, 64, "(%d,%d) %s", x, y, "RBUTTON_UP");
    break;
  case CV_EVENT_MBUTTONUP:
    snprintf (str, 64, "(%d,%d) %s", x, y, "MBUTTON_UP");
    break;
  case CV_EVENT_LBUTTONDBLCLK:
    snprintf (str, 64, "(%d,%d) %s", x, y, "LBUTTON_DOUBLE_CLICK");
    break;
  case CV_EVENT_RBUTTONDBLCLK:
    snprintf (str, 64, "(%d,%d) %s", x, y, "RBUTTON_DOUBLE_CLICK");
    break;
  case CV_EVENT_MBUTTONDBLCLK:
    snprintf (str, 64, "(%d,%d) %s", x, y, "MBUTTON_DOUBLE_CLICK");
    break;
  }

  // (5)マウスボタン，修飾キーを取得
  if (flags & CV_EVENT_FLAG_LBUTTON)
    strncat (str, " + LBUTTON", 64);
  if (flags & CV_EVENT_FLAG_RBUTTON)
    strncat (str, " + RBUTTON", 64);
  if (flags & CV_EVENT_FLAG_MBUTTON)
    strncat (str, " + MBUTTON", 64);
  if (flags & CV_EVENT_FLAG_CTRLKEY)
    strncat (str, " + CTRL", 64);
  if (flags & CV_EVENT_FLAG_SHIFTKEY)
    strncat (str, " + SHIFT", 64);
  if (flags & CV_EVENT_FLAG_ALTKEY)
    strncat (str, " + ALT", 64);

  // (6)マウス座標，イベント，修飾キーなどを画像に描画，表示
  if (line > max_line) {
    cvGetRectSubPix (img, img, cvPoint2D32f (320 - 0.5, 240 - 0.5 + h));
    cvPutText (img, str, cvPoint (w, 20 + h * max_line), &font, CV_RGB (0, 200, 100));
  }
  else {
    cvPutText (img, str, cvPoint (w, 20 + h * line), &font, CV_RGB (0, 200, 100));
  }
  line++;
  cvShowImage ("Image", img);
}
